src/constraints.jl

```diff
@@ -0,0 +1,645 @@
+    structure,
+    name::String,
+    tag::DisjunctConstraint
+)
+    dref = _disjunction(_error, model, structure, name)
+    obj = JuMP.constraint_object(dref)
+    return JuMP.add_constraint(model, _DisjunctConstraint(obj, tag.indicator), name)
+end
+
+# General fallback for additional arguments
+function _disjunction(
+    _error::Function,
+    model::JuMP.Model, # TODO: generalize to AbstractModel
+    structure,
+    name::String,
+    extra...
+)
+    for arg in extra
+        _error("Unrecognized argument `$arg`.")
+    end
+end
+
+"""
+    disjunction(
+        model::JuMP.Model, 
+        structure, 
+        name::String = ""; 
+        nested::Bool = false,
+        indicator::Union{LogicalVariableRef, Nothing} = nothing
+    )
+
+Function to add a [`Disjunction`](@ref) to a [`GDPModel`](@ref). 
+If no indicator is passed, an anonymous [`LogicalVariable`](@ref) is created.
+"""
+function disjunction(
+    model::JuMP.Model, 
+    structure, 
+    name::String = ""; 
+    nested::Bool = false,
+    indicator::Union{LogicalVariableRef, Nothing} = nothing
+) # TODO add kw argument to build exactly 1 constraint
+    if !isnothing(indicator)
+        return _disjunction(error, model, structure, name, DisjunctConstraint(indicator))
+    elseif nested
+        return _disjunction(error, model, structure, name, DisjunctConstraint)
+    else
+        return _disjunction(error, model, structure, name)
+    end
+end
+
+################################################################################
+#                              LOGICAL CONSTRAINTS
+################################################################################
+# _first_order_ops = (
+#     (:Ξ, :exactly),
+#     (:Λ, :atmost), 
+#     (:Γ, :atleast)
+# )
+# for (ops, set) in zip(_first_order_ops, [Exactly, AtMost, AtLeast])
+#     for op in ops
+#         function JuMP.parse_constraint_call(_error::Function, ::Bool, ::Val{op}, val, lvec)
+#             build_code = :(JuMP.build_constraint($(_error), $(esc(lvec)), $set($(esc(val)))))
+#             return :(), build_code
+#         end
+#     end
+# end
+"""
+    function JuMP.parse_constraint_call(
+        _error::Function, 
+        is_vectorized::Bool, 
+        ::Val{op}, 
+        lhs, 
+        rhs
+    )
+
+Extend `JuMP.parse_constraint_call` to avoid order of precedence when 
+using the logical operators ⇔, ⇒, or <--> when building a 
+logical proposition constraint.
+"""
+function JuMP.parse_constraint_call(
+    _error::Function, 
+    is_vectorized::Bool, 
+    val_op::Union{Val{:⇒},Val{:⇔},Val{:(<-->)}}, 
+    lhs, 
+    rhs
+)
+    op = val_op == Val{:⇒} ? :⇒ :
+            val_op == Val{:⇔} ? :⇔ : :(<-->)
+    error_msg = 
+        "The MOI set in the logical constraint $lhs $op $rhs was not identified. " *
+        "The set `MOI.EqualTo{Bool}(value)` should be preceded by `==` or `in`."
+    if rhs.args[1] in (:in, :(==))
+        rhs0 = rhs.args[2]
```

Comment:
A few things here:
- You need to use `_error` instead of `error` to get a nice macro error message.
- The accepted syntax should only be `[logical_expr] in set` where `set` is a `MOI.EqualTo` or `[logical_expr] == true` (or `false`). However, it currently can mix the `in` and `==` operators between these two syntaxes.
- The user might also use `∈` instead of `in`.
- You can only check the type of AST arguments if they are explicit numbers or booleans, otherwise they cannot be determined until runtime. Thus, the  `rhs.args[3] isa` checks will not work as you intend. In reality the acceptable forms of `rhs.args[3]` could include `Bool`, `Symbol` (some Julia variable was given that points to a set or boolean), or `Expr` where the head type is `call` and the first argument is `:(MOI.EqualTo)`. To properly handle the dispatch, you'll need some some function to put in the parse or build code that will do appropriate things at runtime (i.e., when the code is executed).